### DISCUSSION TITLE
How to best usage of GPU memory Usages and Utilization during ASR training

### DISCUSSION THREAD

**Discussion by mehadi92** [NVIDIA-NeMo/NeMo#7474] [Q&A]
Hi,

I'm training an ASR model in a GCP `a2-megagpu-8g` (8 A100 40G) GPU. Here is `nvidia-smi` output
![Screenshot from 2023-09-21 21-15-53](https://github.com/NVIDIA/NeMo/assets/104126711/ef1633d3-0d19-4831-9d34-312bc787c596)

In the above, we can see that `Memory-Usage` is not even. GPU 0 uses high memory but other 1 to 7 GPUs are using very low memory. Sometimes the `Volatile GPU-Util` becomes `0%`. 

In my training setup, I'm using data bucketing with 8 buckets with `bucketing_batch_size = 12`, `num_worker=4`. For faster read-write I'm using [Extreme Persistent Disk](https://cloud.google.com/compute/docs/disks)

To load the model I'm using
```python
model = nemo_asr.models.EncDecCTCModelBPE.restore_from(pretrained_model_path, map_location='cuda')
```
- Should I use `map_location='cpu'` to save some GPU memory in the `GPU 0`? Does it slow down the training If I load my base model into the CPU?
- Should I increase the `bucketing_batch_size` value? I try with 16 instead of 12. But It gives me **CUDA out of memory**  error after some steps complete

Please let me know If you need any other information

Thanks

**Comment by mehadi92** [comment]
Here is an `nvidia-smi` output where the GPU Util become 0
![Screenshot from 2023-09-21 21-39-46](https://github.com/NVIDIA/NeMo/assets/104126711/af7070e1-1a97-4fa1-9797-a655fc34a50d)

**Comment by titu1994** [comment]
You can provide a list of bucketing batch sizes for adaptive bucketing - https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/asr/datasets.html#bucketing-datasets

GPU will always require slightly more memory, but not this much. This is happening because the bucketing batch size is too large for long seq and too small for short seq. 

For validation step, it is fine if GPU is not used cause some metrics have to be calculated on CPU. You can disable log_predictions in config to prevent CPU calculation of transcriptional during training.

**Comment by mehadi92** [comment]
@titu1994, 
If I set `bucketing_batch_size=12` and `batch_size=1` then I'm getting this
```
[NeMo I 2023-09-22 09:07:43 audio_to_text_dataset:750] Batch bucketing is enabled for 8 buckets with adaptive batch sizes of [96, 84, 72, 60, 48, 36, 24, 12]!
```
Do I need to explicitly pass the batch size list? Instead of just setting `bucketing_batch_size=12`

My training data min audio length is 1.5 sec and max audio length is 16.0 I create 8 buckets with 128 shades
How do I find the optimal batch size? In the above setup, I was tracking the GPU Memory Usage For sum steps the GPU was almost full, Here is some stats,
![Screenshot from 2023-09-22 21-42-10](https://github.com/NVIDIA/NeMo/assets/104126711/f4b601a6-0a6d-47a8-a96b-62ac8aae14b1)

Is the batch size 12 too large for 14.18 to 16.0 ( it will be in bucket 8) sec audio in A100 40 G GPU?
Is the batch size 96 too low for 1.5 to 3.31 seconds of audio?

How do I find the optimal bucketing batch size for each 8 buckets?

**Reply by titu1994** [reply]
GPU utilization is not the same as memory. I dunno what the difference between memory.used and utilization.memory. If the first one is the actual memory used, no need to write a list. It even says memory.free = 12 GB so perhaps the first one isn't correct ?

Your utilization.memory on multiple GPU is 57% so maybe try manually pass your list. You will need to experiment with the list of sizes to see how you can max out your GPU memory on each GPU so as not to oom. 

The smallest bucket can have larger batch sizes usually than what we calculate as a good estimate, but the largest bucket should stick to 12.

**Comment by mehadi92** [comment]
@titu1994 
I used this script to write the above CSV file, I Think this script could not write the correct value. The `memory.free + memory.used` must be equal to `memory.total` 
This CSV file seems more accurate [gpu_memory_usage_8gpu_run5.csv](https://github.com/NVIDIA/NeMo/files/12707683/gpu_memory_usage_8gpu_run5.csv) This training is run on 8 A100 GPU for 27 epoch

```sh
while true; do
  nvidia-smi --query-gpu=timestamp,name,memory.total,memory.used,memory.free,utilization.gpu,utilization.memory --format=csv,noheader,nounits >> $logfile
  sleep 20
done
```

If found this docs from `nvidia-smi` tool
```sh
"memory.used"
Total memory allocated by active contexts.
"memory.free"
Total free memory.
"utilization.memory"
Percent of time over the past sample period during which global (device) memory was being read or written
```
- Here is WanDB GPU Utilization
![W B Chart 9_24_2023, 9_35_50 AM](https://github.com/NVIDIA/NeMo/assets/104126711/a5c27b93-af54-4efc-9300-278cf5537e67)
- GPU Memory allocation. We can see there is some JUMPS  in Memory allocation
![W B Chart 9_24_2023, 9_34_36 AM](https://github.com/NVIDIA/NeMo/assets/104126711/318267ee-2b55-4576-9ca8-ca58034180ec)


Thanks for your guidelines. I will try manually set the batch size for each bucket

**Reply by titu1994** [reply]
Interesting, make sure to set benchmarking to false to try to keep gpu memory stable. There is some memory you could safely increase like the 50-70 range buckets up to 80s or 85
